### PR TITLE
utils: replace boost::date_time usage with c++20 std::chrono

### PR DESCRIPTION
**Pull request by theuni**
This would be a very straightforward and uncontroversial change if not for the sad state of std lib implementations which implement this functionality.

As of now, `std::chrono::parse` and friends (c++20 additions) are only available as of gcc 14 and (unreleased) clang 19.

Formatting in chrono is quite useful, so I don't think it makes sense to wait years until we require those compilers.

Instead of waiting around, this PR takes the sledgehammer approach of adding a fully-conformant implementation of the new c++20 chrono features (from the author of `std::chrono` himself). More info available here: https://howardhinnant.github.io/date/date.html

A convenience header adds the `chrono_parse` namespace, which uses the std lib if possible, and `date.h` otherwise.

This allows us to get rid of our last use of `boost::date_time`, and we can begin using more modern c++20-isms as well.

### CODE REVIEW

**Comment by theuni** [comment]
This was noticed while working on #30434, which requires us to install `boost::date_time`. Though vendoring `date.h` here may seem heavy, keep in mind that the status quo is a forward-incompatible mess of boost headers :)

**Comment by hebasto** [comment]
> This allows us to get rid of our last use of `boost::date_time`

Concept ACK on that.

File: src/wallet/rpc/util.cpp [hunk 1]
```diff
@@ -7,31 +7,38 @@
...
+    std::istringstream in{str};
+    std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds> tp;
+    in >> chrono_parse::parse("%Y-%m-%dT%H:%M:%SZ", tp);
+    if (in.fail() || tp < epoch)
         return 0;
-    return (ptime - epoch).total_seconds();
+
+    // Guard against broken implementations which can't handle very old time
+    // values. Speficically, libstdc++ (as of v14) is broken at least on x86_64
+    // with times before 1677-09-21T00:12:44Z .
```
[the remark this answers is not in the document]
Comment (theuni) [reply src/wallet/rpc/util.cpp:36]: Specifically the comparison against the epoch (0) fails. So, if the timestamp is old enough, this should fail but doesn't:
```c++
assert tp < epoch
```

You can play around here: https://godbolt.org/z/G8r837oGf

I'll add more specific info. I should probably submit a bug report upstream as well.

**Comment by theuni** [comment]
> > Formatting in chrono is quite useful, so I don't think it makes sense to wait years until we require those compilers.
> 
> Can you give some examples where it would be useful? Right now it is only used in a single place in the wallet.

I was assuming that it'd be useful in other places where it's simply cumbersome to deal with these conversions like logging. Or as a simplification for `FormatISO8601DateTime` in `util/time.cpp`. But sure, if we're not going to end up using any more of this formatting in the future, I agree it's not worth bothering with for this single function.


> Is the boost header footprint larger than this header? 

Heh. I modified #30434 to only install the headers required by boost::date_time (and its dependencies, recursively). The result:

```bash
cory@nowhere:19:48:40:~/dev/bitcoin (cmake-boost-depends)
 $ find depends/x86_64-pc-linux-gnu/include/boost/ -type f -name "*.hpp" | wc -l
4268
```

I won't even bother adding up the lines of code, but I'm quite sure it's less than 8k :)


> Also, by adding this header the locale linter alerts, so I wonder if the new stuff changes the behavior for users, or if the linter just alerts about unused code.

Hmm, will have a look at this.

> Finally, `ParseISO8601DateTime` can be done in much less than 8k lines of C++ code, so I wonder if just the required parts can be extracted?

I started down this rabbit hole intending to simply rewrite this function, but ended up afraid of introducing bugs due to the locale footguns. Maybe I'm over-complicating it?

I think if you look at this as: "use c++20 functionality with a bridge for compilers which don't support it yet", this PR makes perfect sense. But given the size of that bridge, I agree it's maybe not worth it.

I suppose I was _also_ curious if anyone else was waiting on this functionality for other code.

**Comment by theuni** [comment]
@maflcko All yours :)

**Pull request closed without merging**